src/file.ts

```diff
@@ -2288,18 +2293,69 @@ class File extends ServiceObject<File> {
     const expiresInSeconds =
         Math.round(expiresInMSeconds / 1000);  // The API expects seconds.
 
-    const config: GetSignedUrlConfigInternal = Object.assign({}, cfg);
-
-    config.action = ({
+    const verb = ({
       read: 'GET',
       write: 'PUT',
       delete: 'DELETE',
       resumable: 'POST',
-    } as {[index: string]: string})[config.action];
+    } as {[index: string]: string})[cfg.action];
 
     const name = encodeURIComponent(this.name);
-    config.resource = '/' + this.bucket.name + '/' + name;
+    const resource = `/${this.bucket.name}/${name}`;
+
+    const version = cfg.version || DEFAULT_SIGNING_VERSION;
+
+    const config: GetSignedUrlConfigInternal = Object.assign({}, cfg, {
+      action: verb,
+      expiration: expiresInSeconds,
+      resource,
+      name,
+    });
 
+    let promise: Promise<SignedUrlQuery>;
+    if (version === 'v2') {
+      promise = this.getSignedUrlV2(config);
+    } else if (version === 'v4') {
+      promise = this.getSignedUrlV4(config);
+    } else {
+      throw new Error(`Invalid signed URL version: ${
+          version}. Supported versions are 'v2' and 'v4'.`);
+    }
+
+    promise
+        .then((query) => {
+          if (typeof config.responseType === 'string') {
+            query['response-content-type'] = config.responseType!;
+          }
+
+          if (typeof config.promptSaveAs === 'string') {
+            query['response-content-disposition'] =
+                'attachment; filename="' + config.promptSaveAs + '"';
+          }
+          if (typeof config.responseDisposition === 'string') {
+            query['response-content-disposition'] = config.responseDisposition!;
+          }
+
+          if (this.generation) {
+            query.generation = this.generation.toString();
+          }
+
+          const parsedHost =
+              url.parse(config.cname || STORAGE_DOWNLOAD_BASE_URL);
```

Comment:
I believe `url.parse` is actually deprecated:
https://nodejs.org/docs/latest/api/url.html#url_legacy_url_api

Reply:
Switched to using the `query-string` package.

The built-in `URLSearchParams` incorrectly encodes `;` character in `response-content-disposition` header as separate query params.